### DISCUSSION TITLE
how to use kc_idp_hint parameter through proxy

### DISCUSSION THREAD

**Discussion by KDMRyanT** [keycloak/keycloak#15518] [Q&A] [answered]
Hello,

I have a client setup in a keycloak realm with two IdPs attached in the realm.  This keycloak server is in behind a proxy that is for site specific urls for customers to our application.  What I was thinking keycloak can do is normally when a openid connect client application is connected to keycloak, to get to the application login page:

<KEYCLOAK_HOST>/auth/realms/<YOUR_REALM_NAME>/protocol/openid-connect/auth?client_id=<YOUR_CLIENT_ID>&redirect_uri=<THE_REDIRECT_URI>&response_type=code&scope=openid&kc_idp_hint=<the_name_that_you_give_to_the_idp>

so for example the login page which displays Username/Password and my IdP selection buttons, if its client_id in keycloak was called "myapp" and the IdPs are named idp1 and idp2 in the realm of mycompany, I should be able to pass from apache to talk to my keycloak host, I can setup rewrites or proxy to say the following:

To go directly to application and then redirect to idp1:
https://mykeycloakhost.mydomain.com:8443/auth/realms/mycompany/protocol/openid-connect/auth?client_id=myapp&redirect_uri=https%3A%2F%2Fmyapplication.mydomain.com%3A8443%2Fmyapp&response_type=code&scope=openid&kc_idp_hint=idp1

To go directly to application and then redirect to idp2:
https://mykeycloakhost.mydomain.com:8443/auth/realms/mycompany/protocol/openid-connect/auth?client_id=myapp&redirect_uri=https%3A%2F%2Fmyapplication.mydomain.com%3A8443%2Fmyapp&response_type=code&scope=openid&kc_idp_hint=idp2

Then for the call to idp1, that could be placed on the proxy as https://customer-name.mydomain.com/myapp and because that is rewritten on the proxy to the keycloak url with kc_idp_hint specified, this would work?

Same thing then for the call to idp2, that could be blaced on the proxy as https://customer-name2.mydomain.com/myapp and because I have that url proxied and rewritten to have the correct kc_idp_hint provided - this would appear to the customer seemlessly?

I'm just wondering if this is at possible or this is how it should be done.  How does using apache openid connect mod change this?  Can you use openid connect mod for apache or is this simply a mod_proxy situation?

Please advise.
Thanks

**Comment by KDMRyanT** [comment]
Hello,
For testing this - I have setup a docker keycloak instance - created realm and client application configuration.  For the client application - I'm using docker wildfly with keycloak-quickstarts app-authz-jee-vanilla application.  WIth I'm sure full knowledge on how that quickstart application works... I'm activating the application - https://wildfly.mykdm.dev.local:8446/app-authz-jee-vanilla/?kc_idp_hint=idp1 .  On doing this activation, I'm still redirected to the main keycloak login page.  I know the keycloak wildfly oidc adapter is working as on keycloak login page - I can select one of the Idps - like idp1 above in url call and I'm redirected to the IdP.  Once I login to the IdP with a valid user - this user is then checked against user federation setup in keycloak server.  With a valid user signed in, I'm redirected back to the wildfly jboss server with displayed information on the logged in user - Welcome! Logout Button - Your Permissions are: Resesource: Default Resource ID: <the resource id>.

Just looking for guidance on this - documentation clearly states that:

Client-suggested Identity Provider

OIDC applications can bypass the Keycloak login page by hinting at the identity provider they want to use. You can enable this by setting the kc_idp_hint query parameter in the Authorization Code Flow authorization endpoint.

With Keycloak OIDC client adapters, you can specify this query parameter when you access a secured resource in the application.

For example:

GET /myapplication.com?kc_idp_hint=facebook HTTP/1.1
Host: localhost:8080

In this case, your realm must have an identity provider with a facebook alias. If this provider does not exist, the login form is displayed.

If continuing with the example above with app-authz-vanilla - and an idp alias of facebook - wouldn't https://wildfly.mykdm.dev.local:8446/app-authz-vanilla?kc_idp_hint=facebook

Am I interpreting this incorrectly???

Thanks
Ryan

**Comment by KDMRyanT** [comment, marked as answer]
Hello,
I believe I have this figured out myself.  Basically you have to have the Identity provider redirector there in a authorization flow AND the config option alias has to be set even if left blank.  Now above if I did have facebook Idp configured - going to my application with https://wildfly.mykdm.dev.local:8446/app-authz-vanilla?kc_idp_hint=facebook works.  If you want to make that there has to be a kc_idp_hint always - set the authenticator to Required in the authentication flow... now the application cannot be accessed without kc_idp_hint passed.

**Comment by akshayrrao-Sclera** [comment]
Hello,
I configured following above instruction and used this url : http://localhost:8080/realms/master/protocol/openid-connect/auth?client_id=security-admin-console&redirect_uri=http://localhost:8080/admin/master/console/&response_type=code&scope=openid&kc_idp_hint=idp1 
 but its not working any idea what might be the issue? 

![image](https://github.com/keycloak/keycloak/assets/99779446/5bfc76c6-58c9-41c7-9031-e2db25f74605)

**Reply by danielFesenmeyer** [reply]
I guess people might be able to help you if you could describe what exactly is not working.
Interesting information would be:

- Is an error screen displayed, and if yes, which error message is shown
- If you are not redirected to the IDP, which screen is shown instead?
- Are there any related entries in the logs?
